### PR TITLE
feat(runjob):  add support for cloud_sql_instance

### LINKED ISSUE #1307: How to configure Cloud SQL instance for a RunJob ?
### Describe your question

Hi,
Is it possible to configure Cloud SQL instances for a `RunJob`?
It is [possible for `RunService`](https://cloud.google.com/config-connector/docs/reference/resource-docs/run/runservice#run_service_sql) but I don't see an equivalent for `RunJob`.
There is a mention of SQL instances for `template.template.containers[].volumeMounts[].mountPath` but I don't understand how to specify the instance.
Thank you

### PR DESCRIPTION
**Pull request by CyberHippo**
### Change description

Add support for [`cloud_sql_instance`](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/cloud_run_v2_job#cloud_sql_instance) for Cloud Run Jobs.

Fixes #1307

### Tests you have done

- [x] Run `make ready-pr` to ensure this PR is ready for review.
- [ ] Perform necessary E2E testing for changed resources.
- [ ] go test -v -tags=integration ./config/tests/samples/create -test.run TestAll -run-tests runjob

### CODE REVIEW

**Comment by CyberHippo** [comment]
Hi, @acpana / @anhdle-sso / @barney-s / @cheftako / @diviner524 / @gemmahou / @jasonvigil / @jiefenghe / @jingyih / @justinsb / @maqiuyujoyce / @roycaihw / @yuwenma / @ziyue-101 /@xiaoweim sorry for tagging you directly, could you please take a look at my PR. This would greatly improve the usability of `RunJob` for us. Thanks !

**Comment by gemmahou** [comment]
Hi @CyberHippo would you mind collaborating more on your issue https://github.com/GoogleCloudPlatform/k8s-config-connector/issues/1307? Could you help us understand the business impact and the priority to support this feature? Is there a KCC/Google internal feature request associate to it? Thanks!

**Comment by CyberHippo** [comment]
Hi @gemmahou , thank you for taking a look!

> would you mind collaborating more on your issue https://github.com/GoogleCloudPlatform/k8s-config-connector/issues/1307? 

Sure, let me know what information you need.

> Could you help us understand the business impact and the priority to support this feature? 

We are deploying compute heavy workloads using Cloud Run Jobs and these workloads need to access our Cloud SQL instances. As of now we deploy these Run Jobs using KCC but the lack of Cloud SQL instance support means that we need to create and use Cloud SQL client certificates instead of relying on the SQL socket. It seems that Cloud Run Jobs already support this feature and that it only needs to be added in the RunJob specs for it to work. Please note that KCC Run Service supports Cloud SQL instances.   

> Is there a KCC/Google internal feature request associate to it?

Not that I am aware of but I could open one if needed. I'm not familiar with KCC dev cycle.

**Comment by CyberHippo** [comment]
Friendly ping @gemmahou

**Comment by gemmahou** [comment]
> Friendly ping @gemmahou

Here's the guide for asking a new feature support: https://g3doc.corp.google.com/cloud/config/configconnector/g3doc/customer-engagement.md#recommended-support-channels

We suggest open a GH or Buganizier issue, describe the stakeholder, business impact, and assign to KCC team member or our PM Nic(nslattery). That will help us manage and track feature requests. 

cc: @yuwenma

File: config/servicemappings/run.yaml [hunk 1]
```diff
@@ -84,6 +83,15 @@ spec:
         kind: IAMServiceAccount
         version: v1beta1
         group: iam.cnrm.cloud.google.com
+    - tfField: template.template.volumes.cloud_sql_instance.instances
+      description: |-
+        The Cloud SQL instance connection names, as can be found in https://console.cloud.google.com/sql/instances. Visit https://cloud.google.com/sql/docs/mysql/connect-run for more information on how to connect Cloud SQL and Cloud Run. Format: {project}:{location}:{instance}
+      key: instances
```
Comment (gemmahou) [inline config/servicemappings/run.yaml:89]: Should be `instancesRef` per the naming convention: https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/README.ConfigureResourceReferences.md#configure-reference-resource-in-the-service-mappings, point 3

**Comment by gemmahou** [comment]
We require test coverage for this new field reference: https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/README.NewResourceFromTerraform.md#test-your-changes

Here are some existing test cases for your reference:
Dynamic test: https://github.com/GoogleCloudPlatform/k8s-config-connector/tree/master/pkg/test/resourcefixture/testdata/basic/run/v1beta1/runjob
Samples test: https://github.com/GoogleCloudPlatform/k8s-config-connector/tree/master/config/samples/resources/runjob

None of above tests covered the CloudSQLInstance added in this PR, could you add a new test case(i.e. jobwithsqlinstance), include the field reference(CloudSQLInstance) you add, run the dynamic and samples test to make sure it works as expected?

File: config/servicemappings/run.yaml (hunk 1 above)
Comment (maqiuyujoyce) [reply config/servicemappings/run.yaml:89]: I think for a list-typed reference field, we should use name `instanceRefs` instead.

Comment (CyberHippo) [reply config/servicemappings/run.yaml]: Hi, I used `instances` to be coherent with the [instance field of Run Services](https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/config/servicemappings/run.yaml#L204-L212) but I changed it to `instanceRefs` in https://github.com/GoogleCloudPlatform/k8s-config-connector/pull/1688/commits/249d2177884671256729fe7d315c97559d5df31a

**Comment by CyberHippo** [comment]
Hi @gemmahou , as suggested I added dynamic tests for Run job with SQL deps (please see https://github.com/GoogleCloudPlatform/k8s-config-connector/pull/1688/commits/f13a8dea1e01731cdafe2145a6606494cefb6919).

**Comment by gemmahou** [comment]
@CyberHippo triggered the presubmit test. Could you let us know who is the stakeholder/customer for this feature request? Thanks!

**Comment by CyberHippo** [comment]
> @CyberHippo triggered the presubmit test. Could you let us know who is the stakeholder/customer for this feature request? Thanks!

Hi @gemmahou, thanks for running the presubmit job ! The customer would be the company [elmy](https://elmy.fr/).

File: crds/run_v1beta1_runjob.yaml [hunk 2]
```diff
@@ -602,6 +602,49 @@ spec:
                         description: A list of Volumes to make available to containers.
                         items:
                           properties:
+                            cloudSqlInstance:
+                              description: For Cloud SQL volumes, contains the specific
+                                instances that should be mounted. Visit https://cloud.google.com/sql/docs/mysql/connect-run
+                                for more information on how to connect Cloud SQL and Cloud
+                                Run.
+                              properties:
+                                instances:
```
Comment (gemmahou) [inline crds/run_v1beta1_runjob.yaml:611]: IIUC this file is generated and field name should be updated to `instanceRefs` as well?

**Review by gemmahou — COMMENTED** [review]
Looks good to me, thanks for adding the test! Only one comment regarding the field name.

File: crds/run_v1beta1_runjob.yaml (hunk 2 above)
Comment (CyberHippo) [reply crds/run_v1beta1_runjob.yaml]: Good catch! I missed it running `make manifests`, fixed with https://github.com/GoogleCloudPlatform/k8s-config-connector/pull/1688/commits/26dbc1818468f9b6ea10c7f6a248bf0b07ed1f0d

**Pull request merged**